playback/src/audio_backend/alsa.rs

```diff
@@ -1,14 +1,81 @@
 use super::{Open, Sink};
-use alsa::{Access, Format, Mode, Stream, PCM};
+use alsa::device_name::HintIter;
+use alsa::pcm::{Access, Format, HwParams, PCM};
+use alsa::{Direction, Error, ValueOr};
+use std::env;
+use std::ffi::CString;
 use std::io;
+use std::process::exit;
 
 pub struct AlsaSink(Option<PCM>, String);
 
+fn list_outputs() {
+    for t in &["pcm", "ctl", "rawmidi", "timer", "seq", "hwdep"] {
+        println!("{} devices:", t);
+        let i = HintIter::new(None, &*CString::new(*t).unwrap()).unwrap();
+        for a in i {
+            println!("  {:?}", a)
+        }
+    }
+}
+
+fn open_device(dev_name: &str) -> Result<(PCM), Box<Error>> {
+    let pcm = PCM::new(dev_name, Direction::Playback, false)?;
+    // http://www.linuxjournal.com/article/6735?page=0,1#N0x19ab2890.0x19ba78d8
+    // latency = period_size * periods / (rate * bytes_per_frame)
+    // For 16 Bit stereo data, one frame has a length of four bytes.
+    // 500ms  = buffer_size / (44100 * 4)
+    // buffer_size = 0.5 * 44100 = 22050 frames
+    {
+        // Set hardware parameters: 44100 Hz / Stereo / 16 bit
+        let hwp = HwParams::any(&pcm)?;
+
+        hwp.set_access(Access::RWInterleaved)?;
+        hwp.set_format(Format::s16())?;
+        hwp.set_rate(44100, ValueOr::Nearest)?;
+        hwp.set_channels(2)?;
+        // hwp.set_period_size_near(256, ValueOr::Nearest)?;
+        hwp.set_buffer_size_near(11025 * 2)?; // ~ 0.25 x 2 s latency
+
+        pcm.hw_params(&hwp)?;
+    }
+
+    // Additional software paramters + check
+    if env::var("LIBRESPOT_DEBUG").is_ok() {
+        let hwp = pcm.hw_params_current()?;
+        let swp = pcm.sw_params_current()?;
+        let (bufsize, periodsize) = (hwp.get_buffer_size()?, hwp.get_period_size()?);
+        let periods = hwp.get_periods()?;
+        info!(
+            "periods: {:?} buffer_size: {:?} period_size {:?}",
+            periods, bufsize, periodsize
+        );
+        // Not required now that buffer size is set properly
+        // swp.set_start_threshold(bufsize - periodsize)?;
+        // swp.set_avail_min(periodsize)?;
+        // pcm.sw_params(&swp).unwrap();
+        info!(
+            "Opened audio output {:?} with parameters: {:?}, {:?}",
+            dev_name, hwp, swp
+        );
+    }
+
+    Ok(pcm)
+}
+
 impl Open for AlsaSink {
-    fn open(device: Option<String>) -> AlsaSink {
+   fn open(device: Option<String>) -> AlsaSink {
         info!("Using alsa sink");
 
-        let name = device.unwrap_or("default".to_string());
+        let name = match device.as_ref().map(AsRef::as_ref) {
+            Some("?") => {
+                println!("Listing available alsa outputs");
+                list_outputs();
+                exit(0)
```

Comment:
It doesn't seem right that this library code is calling `exit`. But the portaudio backend is just as wrong in this regard so maybe we can fix that another timer.